qaframework-bdd-tests/src/test/resources/features/refapp-2.x/07-visit/visits.feature

```diff
@@ -1 +1,47 @@
-Feature: Visits management
\ No newline at end of file
+Feature: Visits management
+
+  Background: 
+    Given user logins into the system
+
+  @selenium
+  @visit
+  Scenario: Starting a visit
+    When user clicks on the find patient record app
+    And user searches for the patient record
+    And user clicks on the first patient record
+    And user clicks on the start visit link
+    Then the system starts the patient visit
+
+  @selenium
+  @visit
+  Scenario: Adding a past visit
+    When user clicks on the find patient record app
+    And user searches for the patient record
+    And user clicks on the first patient record
+    And user clicks on the add past visit link
+    Then the system adds the patient past visit
+
+  @selenium
+  @visit
+  Scenario: Merging visits
+    When a patient is initiated with two visits
```

Comment:
is this a step within the user story?

Reply:
Yes it is

Reply:
could you share its equivalent selenium definition?

Reply:
```
    @When("a patient is initiated with two visits")
    public void setUpLoginWithVisits() {
    	new TestData.TestVisit(patient.uuid, TestData.getAVisitType(), getLocationUuid(homePage)).create();
    	new TestData.TestVisit(patient.uuid, TestData.getAVisitType(), getLocationUuid(homePage)).create();
```


Is this what your requesting for?

Reply:
this can be added within the @before method since its a recorded step.

Reply:
- I think these scenarios are too disjointed to have an @before method. Ask me why? Because each of the scenarios wants a different thing. 
- `Starting a visit`: doesnt require a patient visit initiated and so does `Add Past visit`.
- `Merge visits` requires two visits by a patient so the above snippet that I sent works.
- `Ending a visit` requires one visit initiated for one patient.

Now, when I create an @before method and put all those things there, the tests will fail.

Reply:
Look out for my next commit for you to see the behaviour of what am saying...

Reply:
@kdaud am still waiting for your advice on this keeping in mind my [comment here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/348#discussion_r801986529).

Reply:
@mherman22 could you share the error here to take a look?

Reply:
in addition to this [build failure](https://github.com/openmrs/openmrs-contrib-qaframework/runs/5197033225?check_suite_focus=true), you can take a look at [my terminal logs](https://pastebin.com/5DJQ0ygv)

Reply:
The failure is pointing to something different which is not related to the refactoring, could you address the reviews have made on the PR and see how it goes.

Reply:
@kdaud how i wish you pulled this branch and viewed what im talking about via the ui. The logs are not doing me service on this :laughing: . The `start visit` scenario can't be finished because it will throw a message via the ui that `the patient already has active visits and that they should be ended first`. That is where the pointer should be

Reply:
and yet the test somehow goes through the step and jumps onto another scenario but having not done what it is supposed to do

Reply:
are you able to share the screen shoot of the ui?

> The `start visit` scenario can't be finished because it will throw a message via the ui that `the patient already has active visits and that they should be ended first`

Reply:
![Screenshot from 2022-02-15 13-23-46](https://user-images.githubusercontent.com/47120265/154042969-9d5d0eaa-de23-4ac0-ab19-5315bc89e434.png)

Reply:
What happens when you use a demo server?

Reply:
ofcos the steps won't be instantiated by failing to login into the demoserver. something like `Failed to login in VisitSteps.null, trying again...`

Reply:
change the value of [includes.csrftoken](https://github.com/openmrs/openmrs-contrib-qaframework/blob/e3311d4c7c90549da4561b28c00686baf5ca7f27/qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test.properties#L9) to false and run the test again!

Reply:
thanks. it's now able to run the tests. But throws the same as in the screensht at https://github.com/openmrs/openmrs-contrib-qaframework/pull/348#discussion_r806678808

Reply:
![Screenshot from 2022-02-15 13-53-29](https://user-images.githubusercontent.com/47120265/154047852-0e371618-a2ab-42f7-a8ea-2c16ac5a8846.png)

Reply:
take a close look at the ui message, what do you think is the issue?

Reply:
my thinking was that it because in the @before method, the patient is initiated with two visits already using        
```
new TestData.TestVisit(patient.uuid, TestData.getAVisitType(), getLocationUuid(homePage)).create();
    	new TestData.TestVisit(patient.uuid, TestData.getAVisitType(), getLocationUuid(homePage)).create();
```

Reply:
what happens when you get rid of one of the visit?

Reply:
![Screenshot from 2022-02-15 14-14-47](https://user-images.githubusercontent.com/47120265/154051213-8c623d89-9ff9-49fa-9e60-d601a1ab88ca.png)
That is the result in the screenshot

Reply:
so what do you think is the issue?

Reply:
i believe if i removed the remaining visit, it'd pass by doing what it is suppozd to do.

Reply:
so the problem is you can't start a visit when there's an existing visit

Reply:
feel free to test-out things to figure out the better approach to the workflow.

Reply:
alright, i will send in a commit after then